builder/dockerfile/shell_parser.go

```diff
@@ -203,127 +201,122 @@ func (sw *shellWord) processDoubleQuote() (string, error) {
 	//    $ ` " \ <newline>.
 	//  Otherwise it remains literal.
 
-	var result string
+	var result bytes.Buffer
 
 	sw.scanner.Next()
 
+loop:
 	for {
-		ch := sw.scanner.Peek()
-
-		if ch == scanner.EOF {
-			return "", fmt.Errorf("unexpected end of statement while looking for matching double-quote")
-		}
-
-		if ch == '"' {
+		switch sw.scanner.Peek() {
+		case scanner.EOF:
+			return "", errors.New("unexpected end of statement while looking for matching double-quote")
+		case '"':
 			sw.scanner.Next()
-			break
-		}
-
-		if ch == '$' {
-			tmp, err := sw.processDollar()
+			break loop
+		case '$':
+			value, err := sw.processDollar()
 			if err != nil {
 				return "", err
 			}
-			result += tmp
-		} else {
-			ch = sw.scanner.Next()
+			result.WriteString(value)
+		default:
+			ch := sw.scanner.Next()
 			if ch == sw.escapeToken {
 				chNext := sw.scanner.Peek()
-
-				if chNext == scanner.EOF {
+				switch chNext {
```

Comment:
`chNext` is never used elsewhere. This could be `switch sw.scanner.Peek() {` (but it doesn't matter).